pkg/cache/v2/cache.go

```diff
@@ -123,12 +123,13 @@ func (r RawResponse) GetDiscoveryResponse() (*discovery.DiscoveryResponse, error
 	}
 
 	r.isResourceMarshaled = true
```

Comment:
There is a race condition here were for some period of time `r.isResourceMarshaled` is `true` and the value of `r.marshaledResponse` is `nil`. This all needs to go in a mutex or potentially just set `r.isResourceMarshaled` after setting the value of `r.marshaledResponse` and accept that occasionally you will marshal a few times but no always.

Reply:
Also, note that as per #339 this will still not fix the lack of caching since on subsequent invocations the `r.isResourceMarshaled` is always `false`.

Reply:
Good point about the race condition. The server is concurrent, so there could be multiple routines operating on the same cache response. I wonder if mutex is going to lead to a contention (maybe it's best to pre-compute it using Passthrough response). @jessicayuen @jyotimahapatra any thoughts on this?

Reply:
I also opened #340 with a test and more aggressive fix.  @alecholmez noted that there are mutexes being copied around there so we'd need to fix that in mine as he'd done in #335 now.

Reply:
I too would avoid a mutex if possible for this and just live with the potential for the occasional wasted cycles but assuming that you'll marshal everything once precomputing makes much more sense.

Reply:
This makes sense to me. However, I think go might reorder statements and can still cause subtle bugs. I guess using an atomic store/load is also necessary. WDYT?

Reply:
Another possible way could be to remove the dependency on the extra `isResourceMarshaled` variable and base the check on the `nil`ness of `r.marshaledResponse`. I think this could remove the concern about statement reordering.

Reply:
The nil check works for me.  Of course this still doesn’t change the fact that it will always be nil until this method operates on a pointer of a RawMessage not a RawMessage value.

Reply:
i think change order is gool...

Reply:
I'm in favor of the `nil` check with the pointer upgrades, I can make that PR

Reply:
hehe, made the change in mine this morning and never pushed.  #340